src/mesh/replicated_mesh.C

```diff
@@ -1417,5 +1417,151 @@ dof_id_type ReplicatedMesh::n_active_elem () const
+  // use a comparison functor to avoid run-time randomness due to pointers
+  struct boundary_side_compare
+  {
+    bool operator()(const std::pair<const Elem *, unsigned int> & lhs,
+                    const std::pair<const Elem *, unsigned int> & rhs) const
+    {
+      if (lhs.first->id() < rhs.first->id())
+        return true;
+      else if (lhs.first->id() == rhs.first->id())
+      {
+        if (lhs.second < rhs.second)
+          return true;
+      }
+      return false;
+    }
+  };
+  std::set<std::pair<const Elem *, unsigned int>, boundary_side_compare> boundary_elements;
+  for (const auto & elem : active_element_ptr_range())
+    for (auto s : elem->side_index_range())
+      if (elem->neighbor_ptr(s) == nullptr)
+        boundary_elements.insert(std::pair<const Elem *, unsigned int>(elem, s));
+
+  while (!boundary_elements.empty())
+  {
+    // get the first entry as the seed
+    const Elem * eseed = boundary_elements.begin()->first;
+    unsigned int sseed = boundary_elements.begin()->second;
+
+    // get the subdomain ID that these boundary sides attached to
+    subdomain_id_type subdomain_id = subdomains[eseed->id()];
+
+    // start visiting the mesh to find all boundary node with the seed
+    std::vector<Point> bpoints;
+    const Elem * elem = eseed;
+    unsigned int s = sseed;
+    std::vector<unsigned int> local_side_nodes = elem->nodes_on_side(s);
+    while (true)
+    {
+      boundary_elements.erase(std::pair<const Elem *, unsigned int>(elem, s));
+
+      // push all nodes on the side except the last one
+      for (unsigned int i = 0; i < local_side_nodes.size() - 1; ++i)
+        bpoints.push_back(*dynamic_cast<const Point *>(elem->node_ptr(local_side_nodes[i])));
+
+      // use the last node to find next element and side
+      const Node * node = elem->node_ptr(local_side_nodes[local_side_nodes.size() - 1]);
+      std::set<const Elem *> neighbors;
+      elem->find_point_neighbors(*node, neighbors);
+
+      // if only one neighbor is found (itself), this node is a cornor node on boundary
+      if (neighbors.size() != 1)
+        neighbors.erase(elem);
+
+      // found the connecting side
+      bool found = false;
+      for (const auto & neighbor : neighbors)
+      {
+        for (auto ss : neighbor->side_index_range())
+          if (neighbor->neighbor_ptr(ss) == nullptr && !(elem == neighbor && ss == s))
+          {
+            local_side_nodes = neighbor->nodes_on_side(ss);
+            for (auto & nd : local_side_nodes)
+              if (neighbor->node_ptr(nd) == node)
+              {
+                elem = neighbor;
+                s = ss;
+                found = true;
+                break;
+              }
+            if (found)
+              break;
+          }
+      }
+      if (!found)
+        libmesh_error_msg("ERROR: mesh topology error on visiting boundary sides");
+
+      // exit if we reach the starting point
+      if (elem == eseed && s == sseed)
+        break;
```

Comment:
This algorithm provides an ordered boundary points that form an enclosure. The ordering is kind of important. Number of separated boundaries is also a product of this algorithm.